integration_tests/pipelined_memdb_test.go

```diff
@@ -0,0 +1,180 @@
+}
+
+type testPipelinedMemDBSuite struct {
+	suite.Suite
+	cluster testutils.Cluster
+	store   *tikv.KVStore
+}
+
+func (s *testPipelinedMemDBSuite) SetupTest() {
+	if *withTiKV {
+		s.store = NewTestStore(s.T())
+		return
+	}
+
+	client, pdClient, cluster, err := unistore.New("", nil)
+	s.Require().Nil(err)
+
+	unistore.BootstrapWithSingleStore(cluster)
+	s.cluster = cluster
+	store, err := tikv.NewTestTiKVStore(fpClient{Client: &unistoreClientWrapper{client}}, pdClient, nil, nil, 0)
+	s.Require().Nil(err)
+
+	s.store = store
+}
+
+func (s *testPipelinedMemDBSuite) TearDownTest() {
+	s.store.Close()
+}
+
+func (s *testPipelinedMemDBSuite) TestPipelinedAndFlush() {
+	ctx := context.Background()
+	txn, err := s.store.Begin(tikv.WithPipelinedMemDB())
+	s.Nil(err)
+	txn1, err := s.store.Begin()
+	s.Nil(err)
+	s.True(txn.IsPipelined())
+	for i := 0; i < MinFlushKeys; i++ {
+		key := []byte(strconv.Itoa(i))
+		value := make([]byte, MinFlushSize/MinFlushKeys-len(key)+1)
+		txn.Set(key, value)
+		flushed, err := txn.GetMemBuffer().Flush(false)
+		s.Nil(err)
+		if i < MinFlushKeys-1 {
+			s.False(flushed)
+		} else {
+			s.True(flushed)
+		}
+		// txn can always read its own writes
+		val, err := txn.Get(ctx, key)
+		s.Nil(err)
+		s.True(bytes.Equal(value, val))
+		// txn1 cannot see it
+		_, err = txn1.Get(ctx, key)
+		s.True(tikverr.IsErrNotFound(err))
+	}
+
+	// commit the txn, then it's visible to other txns.
+	txn.Commit(ctx)
+
+	txn1, err = s.store.Begin()
+	s.Nil(err)
+	for i := 0; i < MinFlushKeys; i++ {
+		key := []byte(strconv.Itoa(i))
+		expect := make([]byte, MinFlushSize/MinFlushKeys-len(key)+1)
+		val, err := txn1.Get(ctx, key)
+		s.Nil(err)
+		s.True(bytes.Equal(expect, val))
+	}
+}
+
+func (s *testPipelinedMemDBSuite) TestPipelinedMemDBBufferGet() {
+	ctx := context.Background()
+	txn, err := s.store.Begin(tikv.WithPipelinedMemDB())
+	s.Nil(err)
+	for i := 0; i < 100; i++ {
+		key := []byte(strconv.Itoa(i))
+		value := key
+		txn.Set(key, value)
+		flushed, err := txn.GetMemBuffer().Flush(true)
+		s.Nil(err)
+		s.True(flushed)
+	}
+	for i := 0; i < 100; i++ {
+		key := []byte(strconv.Itoa(i))
+		expect := key
+		val, err := txn.GetMemBuffer().Get(ctx, key)
+		s.Nil(err)
+		s.True(bytes.Equal(val, expect))
+	}
+	s.Nil(txn.Rollback())
+}
+
+func (s *testPipelinedMemDBSuite) TestPipelinedFlushBlock() {
+	txn, err := s.store.Begin(tikv.WithPipelinedMemDB())
+	s.Nil(err)
+	txn.Set([]byte("key1"), []byte("value1"))
+
+	s.Nil(failpoint.Enable("tikvclient/beforePipelinedFlush", `pause`))
+	flushed, err := txn.GetMemBuffer().Flush(true)
+	s.Nil(err)
+	s.True(flushed)
+
+	txn.Set([]byte("key2"), []byte("value2"))
+	flushReturned := make(chan struct{})
+	go func() {
+		flushed, err := txn.GetMemBuffer().Flush(true)
+		s.Nil(err)
+		s.True(flushed)
+		close(flushReturned)
+	}()
+
+	oneSec := time.After(time.Second)
+	select {
+	case <-flushReturned:
+		s.Fail("Flush should be blocked")
+	case <-oneSec:
+	}
```

Comment:
When the flush is blocked, need to verify the latter write would be stalled when membuf  is full.

Reply:
Latter write should not come during the flush blcok period.

The application may looks like this, no parallel `Flush` and `Set`:

```
for {
  txn.Set(...)
  txn.GetMemBuffer().Flush(false)
}